lib/perso.ml

```diff
@@ -1216,6 +1239,295 @@ let warning_use_has_parents_before_parent (fname, bp, ep) var r =
   |> !GWPARAM.syslog `LOG_WARNING;
   r
 
+let cousins_t = ref None
+let cousins_dates_t = ref None
+
+let max_l1_l2_aux env =
+  let max_l1 = match get_env "max_anc_level" env with Vint i -> i | _ -> 5 in
+  let max_l2 = match get_env "max_desc_level" env with Vint i -> i | _ -> 5 in
+  (max_l1 - 1, (2 * max_l1) + max_l2 + 20)
+
+let get_min_max_dates conf base l =
+  let dates =
+    let rec loop acc = function
+      | [] -> acc
+      | (ip, _, _, _) :: l -> (
+          let not_dead = get_death (poi base ip) = NotDead in
+          let birth_date, death_date, _ =
+            Gutil.get_birth_death_date (poi base ip)
+          in
+          match (birth_date, death_date) with
+          | Some (Dgreg (b, _)), Some (Dgreg (d, _)) ->
+              let birth =
+                match b.prec with
+                | After | Before | About | Maybe | OrYear _ | YearInt _ -> false
+                | _ -> true
+              in
+              let death =
+                match d.prec with
+                | After | Before | About | Maybe | OrYear _ | YearInt _ -> false
+                | _ -> true
+              in
+              if birth && death then loop (b.year :: d.year :: acc) l
+              else if birth && not death then loop (b.year :: acc) l
+              else if (not birth) && death then loop (d.year :: acc) l
+              else loop acc l
+          | Some (Dgreg (b, _)), _ -> (
+              match b.prec with
+              | After | Before | About | Maybe | OrYear _ | YearInt _ ->
+                  if not_dead then loop (conf.today.year :: acc) l
+                  else loop acc l
+              | _ ->
+                  if not_dead then loop (conf.today.year :: b.year :: acc) l
+                  else loop (b.year :: acc) l)
+          | _, Some (Dgreg (d, _)) -> (
+              match d.prec with
+              | After | Before | About | Maybe | OrYear _ | YearInt _ ->
+                  loop acc l
+              | _ -> loop (d.year :: acc) l)
+          | _, _ ->
+              if not_dead then loop (conf.today.year :: acc) l else loop acc l)
+    in
+    loop [] l
+  in
+  List.sort_uniq compare dates
+
+let rec ascendants base acc l i =
+  match l with
+  | [] -> acc
+  | (ip, _, _, lev) :: l -> (
+      match get_parents (poi base ip) with
+      | Some ifam ->
+          let cpl = foi base ifam in
+          let ifath = get_father cpl in
+          let imoth = get_mother cpl in
+          let acc = [ (ifath, [], ifath, [ List.hd lev + 1 ]) ] @ acc in
+          let acc = [ (imoth, [], imoth, [ List.hd lev + 1 ]) ] @ acc in
```

Comment:
no need for making a temporary list, you can write `let acc = (imoth, [], imoth, [ List.hd lev + 1 ]) :: acc in`